app/scripts/controllers/transactions/index.js

```diff
@@ -1542,124 +1602,12 @@ export default class TransactionController extends EventEmitter {
     txMeta = await this._addTransactionGasDefaults(txMeta);
 
     if ([TransactionType.swap, TransactionType.swapApproval].includes(type)) {
-      txMeta = await this._createSwapsTransaction(swaps, type, txMeta);
+      txMeta = await this.createSwapsTransaction(swaps, type, txMeta);
```

Comment:
As this is now an external function, would it be cleaner to call it `updateSwapsTransaction` as the transaction will still exist with the basic metadata, but we allow external entities to change it?

Reply:
This refactor is really well done, but on reflection I'm wondering how much we're gaining by moving all this logic into the client via a callback.

With this new code, the controller knows about the swap transaction types, the swaps properties as they're defined in the `TransactionMeta` type, plus all the swaps specific data is passed to `addTransaction` via the `swaps` option.

The client gives the swaps data to the controller via the `addTransaction` method, the controller then gives it back to the client via `createSwapsTransaction`, and the client then gives it back to the controller by calling `updateTransactionSwapProperties` where it is added to the metadata.

As it's already optional when calling `addTransaction`, meaning the client doesn't have to provide it, is it simpler to stick with what we have and leave the update logic inside the controller, and continue to fire the `newSwap` and `newSwapApproval` events so the client can do anything extra, such as invoking the `SwapsController` as the extension currently does.

This PR is still necessary regardless, to move the swaps metrics logic out of the controller, plus to add the `TransactionEvent` enum and the `lib/transaction` folder which are both great additions.

Reply:
True, there is no point of back and forth between client and controller, rolled back in https://github.com/MetaMask/metamask-extension/pull/21456/commits/86dfac2fc9acf571f3240e2ea59948f8ed6b4a1f